File: public/code/oscillators__cci.ts
import { Candle, Signal, makeSignal, sma, ema, atr } from "../indicators";

export function cci(c: Candle[]): Signal {
  if (c.length < 40) return makeSignal({ reason: "Insufficient data" });
  const p = 20;
  const tp = c.map((x) => (x.high + x.low + x.close) / 3);
  const smaTP = sma(tp, p);
  const i = c.length - 1, a = atr(c, 14);
  const slice = tp.slice(i - p + 1, i + 1);
  const mean = smaTP[i];
  const md = slice.reduce((s, v) => s + Math.abs(v - mean), 0) / p;
  if (md === 0) return makeSignal({ reason: "Flat" });
  const cciNow = (tp[i] - mean) / (0.015 * md);
  const prevSlice = tp.slice(i - p, i);
  const prevMean = smaTP[i - 1];
  const prevMd = prevSlice.reduce((s, v) => s + Math.abs(v - prevMean), 0) / p;
  const cciPrev = prevMd ? (tp[i - 1] - prevMean) / (0.015 * prevMd) : 0;
  const cur = c[i];
  if (cciPrev < -100 && cciNow > -100) {
    const sl = cur.low - 1.5 * a[i], r = cur.close - sl;
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [cur.close + r * 1.5, cur.close + r * 2.5, cur.close + r * 4], confidence: 0.71, reason: `CCI back above -100 (${cciNow.toFixed(0)})` });
  }
  if (cciPrev > 100 && cciNow < 100) {
    const sl = cur.high + 1.5 * a[i], r = sl - cur.close;
    return makeSignal({ signal: "short", entry: cur.close, stop_loss: sl, take_profit: [cur.close - r * 1.5, cur.close - r * 2.5, cur.close - r * 4], confidence: 0.71, reason: `CCI back below +100 (${cciNow.toFixed(0)})` });
  }
  return makeSignal({ reason: `CCI ${cciNow.toFixed(0)} (no cross)` });
}

File: public/code/oscillators__cci.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def cci(c):
    if len(c) < 40:
        return make_signal(reason="Insufficient data")
    p = 20
    tp = [(x['high'] + x['low'] + x['close']) / 3 for x in c]
    smaTP = sma(tp, p)
    i = len(c) - 1
    a = atr(c, 14)
    slice = tp[i - p + 1: i + 1]
    mean = smaTP[i]
    md = sum(abs(v - mean) for v in slice) / p
    if md == 0:
        return make_signal(reason="Flat")
    cciNow = (tp[i] - mean) / (0.015 * md)
    prevSlice = tp[i - p: i]
    prevMean = smaTP[i - 1]
    prevMd = sum(abs(v - prevMean) for v in prevSlice) / p
    cciPrev = ((tp[i - 1] - prevMean) / (0.015 * prevMd) if prevMd else 0)
    cur = c[i]
    if cciPrev < -100  and  cciNow > -100:
        sl = cur['low'] - 1.5 * a[i]
        r = cur['close'] - sl
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] + r * 1.5, cur['close'] + r * 2.5, cur['close'] + r * 4], confidence=0.71, reason=f"CCI back above -100 ({cciNow})")
    if cciPrev > 100  and  cciNow < 100:
        sl = cur['high'] + 1.5 * a[i]
        r = sl - cur['close']
        return make_signal(signal="short", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] - r * 1.5, cur['close'] - r * 2.5, cur['close'] - r * 4], confidence=0.71, reason=f"CCI back below +100 ({cciNow})")
    return make_signal(reason=f"CCI {cciNow} (no cross)")
